app/controllers/replies_controller.rb

```diff
@@ -297,19 +306,153 @@ def require_edit_permission
+
+  def preview_replies
     @page_title = @post.subject
 
-    @reply = @written # So that editor_setup shows the correct characters
     editor_setup
     render :preview
   end
 
+  def post_replies(new_reply: nil)
+    @multi_replies << new_reply if new_reply.present?
+
+    first_reply = @multi_replies.first
+    begin
+      Reply.transaction { @multi_replies.each(&:save!) }
+    rescue ActiveRecord::RecordInvalid => e
+      errored_reply = @multi_replies.detect { |r| r.errors.present? } || first_reply
+      render_errors(errored_reply, action: 'created', now: true, err: e)
+
+      redirect_to posts_path and return unless errored_reply.post
+      redirect_to post_path(errored_reply.post) and return
+    end
+
+    if @multi_replies.length == 1
+      flash[:success] = "Reply posted."
+    else
+      flash[:success] = "Replies posted."
+    end
+    redirect_to reply_path(first_reply, anchor: "reply-#{first_reply.id}")
+  end
+
+  def editing_multi_reply?
+    # If the list of params is present and the first item on the list has the ID stored, I am editing it
+    @multi_replies_params.present? && (@reply = Reply.find_by_id(@multi_replies_params.first["id"]))
+  end
+
+  def edit_reply(editing_multi_reply, new_multi_reply: nil)
+    begin
+      Reply.transaction do
+        edit_multi_replies(new_multi_reply) if editing_multi_reply
+
+        @reply.save!
+      end
+    rescue ActiveRecord::RecordInvalid => e
+      if @multi_replies.blank?
+        render_errors(@reply, action: 'updated', now: true, err: e)
+      else
+        errored_reply = @multi_replies.detect { |r| r.errors.present? } || @reply
+        render_errors(errored_reply, action: 'updated', now: true, err: e)
+      end
+
+      @post ||= @reply.post
+      @audits = { @reply.id => @post.audits.count }
+      editor_setup
+      render :edit
+    else
+      flash[:success] = "Reply updated."
+      redirect_to reply_path(@reply, anchor: "reply-#{@reply.id}")
+    end
+  end
+
+  def edit_multi_replies(new_multi_reply)
+    # Add new replies after the one being edited and before the next
+    if new_multi_reply.present?
+      # Including the reply in the text editor, not just the ones in the JSON
+      @multi_replies << new_multi_reply
+      @multi_replies_params << permitted_params
+    end
+    original_reply = @reply.dup
+
+    # Modify the original reply with the parameters of the first reply in the JSON
+    @reply.assign_attributes(@multi_replies_params.shift)
+    @multi_replies.shift
+
+    # Check whether there are any further replies beyond the very first one
+    num_new_replies = @multi_replies_params.length
+    return if num_new_replies == 0
+
+    # Reorder the replies after this one
+    original_reply_order = @reply.order
+    following_replies = @reply.post.replies.where("reply_order > ?", original_reply_order)
+    following_replies.update_all(["reply_order = reply_order + ?", num_new_replies]) # rubocop:disable Rails/SkipsModelValidations
```

Comment:
Surely you don't want to do that until you actually post it? This should probably go inside the transaction for that

Reply:
That is inside that transaction, that function is only called from within the transaction

Reply:
Oh. sneaky. Okay that makes... sense?